File: src/lib/acronym.ts
// WHY: Mirrors backend logic in crates/yaml-db/src/acronym.rs
export function generateAcronym(name: string, displayName?: string): string {
  const source = displayName?.includes(' ') ? displayName : name
  const cleaned = source.toUpperCase().replace(/-/g, ' ').replace(/_/g, ' ')
  const words = cleaned.split(/\s+/).filter(w => w.length > 0)

  if (words.length === 0) return 'UNKN'

  if (words.length === 1) {
    const word = words[0]
    if (word.length >= 4) {
      const consonants = word
        .split('')
        .filter(c => !'AEIOU'.includes(c))
        .slice(0, 4)
        .join('')

      if (consonants.length >= 4) {
        return consonants
      }
      return word.slice(0, 4)
    }
    const lastChar = word[word.length - 1]
    return (word + lastChar.repeat(4 - word.length)).toUpperCase()
  }

  if (words.length === 2) {
    return (words[0].slice(0, 2) + words[1].slice(0, 2)).toUpperCase()
  }

  if (words.length === 3) {
    return (words[0].slice(0, 2) + words[1][0] + words[2][0]).toUpperCase()
  }

  return words
    .slice(0, 4)
    .map(w => w[0])
    .join('')
    .toUpperCase()
}

export function normalizeFeatureName(name: string): string {
  return name.toLowerCase().trim().replace(/\s+/g, '-').replace(/_+/g, '-')
}

File: crates/yaml-db/src/acronym.rs
/// Validate acronym format: exactly 4 uppercase alphanumeric characters
pub fn validate_acronym_format(acronym: &str) -> Result<(), String> {
    if acronym.len() != 4 {
        return Err(format!(
            "Acronym must be exactly 4 characters, got: {}",
            acronym
        ));
    }

    if !acronym
        .chars()
        .all(|c| c.is_ascii_alphanumeric() && (c.is_ascii_uppercase() || c.is_ascii_digit()))
    {
        return Err("Acronym must contain only uppercase letters and numbers".to_string());
    }

    Ok(())
}

/// Generate 4-letter acronym from name (used only during migration)
/// Examples: "frontend" → "FRNT", "user-profile" → "USPR", "api" → "APIS"
pub fn generate_acronym(name: &str, display_name: &str) -> String {
    let source = if display_name.contains(' ') {
        display_name
    } else {
        name
    };

    let cleaned = source.to_uppercase().replace('-', " ").replace('_', " ");
    let words: Vec<&str> = cleaned.split_whitespace().collect();

    let acronym = match words.len() {
        0 => "UNKN".to_string(),
        1 => {
            let word = words[0];
            if word.len() >= 4 {
                // Extract consonants preferentially, fall back to first 4
                let consonants: String = word
                    .chars()
                    .filter(|c| !"AEIOU".contains(*c))
                    .take(4)
                    .collect();

                if consonants.len() >= 4 {
                    consonants
                } else {
                    word.chars().take(4).collect()
                }
            } else {
                // Pad short words by repeating last letter
                if let Some(last_char) = word.chars().last() {
                    format!(
                        "{}{}",
                        word,
                        last_char.to_string().repeat(4 - word.len())
                    )
                } else {
                    "UNKN".to_string()
                }
            }
        }
        2 => {
            // 2 letters from each word
            format!(
                "{}{}",
                words[0].chars().take(2).collect::<String>(),
                words[1].chars().take(2).collect::<String>()
            )
        }
        3 => {
            // 2 from first, 1 from second, 1 from third
            format!(
                "{}{}{}",
                words[0].chars().take(2).collect::<String>(),
                words[1].chars().next().unwrap_or('X'),
                words[2].chars().next().unwrap_or('X')
            )
        }
        _ => {
            // 4+ words: first letter from first 4 words
            words
                .iter()
                .take(4)
                .filter_map(|w| w.chars().next())
                .collect()
        }
    };

    acronym.to_uppercase()
}
